### ISSUE TITLE
Linux Support

### ISSUE THREAD

**Issue by SamHSmith** [x13pixels/remedybg-issues#176]
I would love to use Remedybg since it seems to be an amazing debugger. However I rarely work on windows as most of my development and general PC use is on Linux. How much of remedy is tied to Windows and would you consider porting it to Linux? I would love to help you do it!

**Comment by x13pixels** [comment]
After looking over the ptrace API the answer is probably a no for now. Maybe there is a less flaky API that is more reliable across multiple versions of the Linux kernel / distributions?

Here is one excerpt from the ptrace documentation:
```
       As of Linux 2.6.38, after the tracer sees the tracee ptrace-stop and
       until it restarts or kills it, the tracee will not run, and will not
       send notifications (except SIGKILL death) to the tracer, even if the
       tracer enters into another waitpid(2) call.

       The kernel behavior described in the previous paragraph causes a
       problem with transparent handling of stopping signals.  If the tracer
       restarts the tracee after group-stop, the stopping signal is effec‐
       tively ignored—the tracee doesn't remain stopped, it runs.  If the
       tracer doesn't restart the tracee before entering into the next
       waitpid(2), future SIGCONT signals will not be reported to the
       tracer; this would cause the SIGCONT signals to have no effect on the
       tracee.

       Since Linux 3.4, there is a method to overcome this problem: instead
       of PTRACE_CONT, a PTRACE_LISTEN command can be used to restart a
       tracee in a way where it does not execute, but waits for a new event
       which it can report via waitpid(2) (such as when it is restarted by a
       SIGCONT).
```

I read this as "after so and so version of Linux the functionality is hosed but it works again in 3.4!". This kind of thing appears all throughout the documentation. It might just be that I am not familiar enough with the API. My apologies if my understanding of ptrace is incorrect.

Then there is the problem with multiple distributions. Even if the API was stable would I have to worry about each of these [distros](https://futurist.se/gldt/wp-content/uploads/12.02/gldt1202.svg) to create a robust product?

And finally, there are less than 1% of people that use Linux (at least according to latest Steam Survey) and less than 1% of those willing to pay any money for software tool...

**Comment by SamHSmith** [comment]
You are right that the api changed a lot up until 3.4. However that was 8 years ago. I would say that almost all desktop users are on kernel 5.0 or later. So you wouldn't have to worry about that at all I don't think.

The answer to your second question is no. The API is stable and the interfaces for creating a gui application are distro agnostic. Ptrace goes directly to the kernel. I assume that your using software rendering in which case nothing changes on windows vs any linux distro. If you are not using software rendering, access to vulkan and opengl are the same across all distros. To interface with the window system you need to communicate to an X server over the X11 protocol. This is standardized across all distros. And even people like myself who are running the new protocol, wayland, can still use X11 apps because we have the compatibility layer called xwayland.

Now normally you don't interface with Xlib directly. You use a toolkit/library like SDL/GLFW/GTK/QT.

And about the money. It's not that Linux users don't like paying for things. It's that they don't like *having* to pay for things. They will happily donate money often out paying what you could have charged for the software had you wanted to sell it to them. However, Linux and the free-software movement has it's roots in broke teenagers who develop software in a garage. Many of us have been in a situation where we didn't have money to pay for tools. But we were able to make things using free-software. Linus Torvalds started working on the linux kernel because he could not afford a desktop OS. And for that reason linux people get very nervous around paywalls. They are however very open to donating money so that the software can be worked on full time.

If you only depend on libc, sys/ptrace.h, and SDL or GLFW I **guarantee** you it will work on every desktop Linux distro out there. And considering you'll be the best debugger the linux world has ever seen I don't think you'd have a hard time finding a considerable amount of customers. Linux developers *are* willing to spend money on tools. Especially since your tool is not made by some big corporation. And you can definatly count every single windows based dev who has to work on linux stuff will come running for remedybg on linux.

Have I succeeded in impacting your thoughts?

**Comment by tdgroot** [comment]
Also interested in seeing Linux support. When that support hits, I'll switch over from gdb guis and buy Remedybg!

As for the market share argument: I think the steam survey is not representative of the Linux community market share in developer space, for obvious reasons.

**Comment by SamHSmith** [comment]
Good point. I would point aswell that google pushing stadia, Linux being the number one server platform, are reasons why a lot of developers and a lot of companies would pay good money for better tools on Linux.

**Comment by jcalabro** [comment]
Hey there! Wanted to leave a note saying that I would also use this on Linux (though I'm already a paying Windows user 😄). I keep ports of my code going on both Windows and Linux, and it would be nice to have the debugger on Linux so that process is smoother. _Huge_ fan of RemedyBG on Windows, and nothing even comes close on Linux.

Beyond just shipping games, I think it would be a boon to have over there as well for all varieties of software! Apart from games, I write a lot of stuff that runs on Linux servers, so it would be really helpful to have for sure. Definitely not a small undertaking to target this though, so totally understood if it's not something that's likely to come down the pipe.

Regardless, thanks for an awesome product!

**Comment by Cthutu** [comment]
Linux will definitely be a bigger player in the video games scene.  I'm in the industry and we're already talking about Steam Deck, and I personally use Pop!_OS that runs most of my games so far.  I've purchased this for Windows and I would love to see this on Linux too.

Also, many game servers are being developed on Linux to support multiplayer games.

**Comment by ryuukk** [comment]
Lot of programs / game server needs to run on linux boxes nowadays

Being able to have a good debugger not only is needed, but the market is quite empty on linux, despite the demand growing year after year

If we fed up with windows, and feel like linux is not quite there, then let's join hands in trying to make linux the best place for devs, i think it is personally, we just need to convince the doubters, let's all do our best, at least we should try, nobody will do it for us

**Comment by SamHSmith** [comment]
That is all true. However I can not say in good faith anymore that I want Linux to become the best place for devs. Because that fundamentally limits how good things could be to kinda shitty. Linux is a bad place to be for more reasons then not having a good debugger. I think the collective effort should go into replacing Windows and Linux instead. That way we could have something truly good instead of less shit.

**Comment by neonmoe** [comment]
I would like to express support for this as well. I would buy RemedyBG in an instant, if I could use it. The only time I've used windows in the past few years has been building executables with msvc in a VM. On Linux, the debugger situation is horrible, basically the only one I've managed to find any value from is the Visual Studio Code CodeLLDB extension, which feels sluggish and doesn't have all the neat features RemedyBG has. There's also the Steam Deck, which could use a proper debugger for devs who want to support the platform. Sorry to just contribute yet another post just begging for this, but I hope this shows that there's many people who would like to buy the program, if it were on Linux.

**Comment by kieranvs** [comment]
The debugger situation on Linux is so awful - I would be happy to pay 10x the current asking price if it supported Linux... :(

**Comment by Szwagi** [comment]
With Steam Deck and systems such as Pop!_OS, I think it'd be amazing to see Remedy working under Linux.

**Comment by Dom324** [comment]
I'm also currently considering buying RemedyBG, but I rarely (only when I'm forced to) use Windows for programming, so a Linux version would be a buy.

**Comment by smslocum** [comment]
Just tacking another vote onto the petition. I'd happily pay for such a tool.

**Comment by RobertLemmens** [comment]
as a paying windows users, I will buy remedy again if that can help towards making linux support happen. I develop cross platform and find myself not wanting to debug in linux 99% of the time.

**Comment by TheGag96** [comment]
Ever since I heard about RemedyBG, I've really wanted to try it out, but nowadays, I run Linux only at home... On Windows at work, I have a limited set of tools, constantly being frustrated that my only choices are the primitiveness of GDB on the command line or abysmal, buggy, crashy GUI wrappers for it written in Python. At the very least, I can say that I would absolutely pay full price for a Linux version. I love buying great software - I've paid for Sublime Text (twice because my license expired lol), Sublime Merge, and Beyond Compare. 

The use of the Steam Survey as reasoning for lack of interest seems very strange to me because - aside from the obvious fact that survey only serves to show the specific demographics for PC gamers - as far as I know, there are a *lot* of people out there developing native software for Linux. Especially given the high amount of bugs in Linux software, the ecosystem could certainly use a strong, sane graphical debugger!!

**Comment by mgarcia-org** [comment]
+1 linux

**Comment by marvhus** [comment]
I would too buy it in an instant if it got Linux support.  
The Linux space really needs some good debuggers.

**Comment by hsnyder** [comment]
+1 request for linux support. I bought RemedyBG and use it on Windows. I love it. I would love to be able to use it on linux too, I would even purchase the Linux version separately.

**Comment by smari** [comment]
+1 on this request. Would love this on Linux.

**Comment by TheEmeraldFalcon** [comment]
Leaving my two cents, I for one would really appreciate something like this, the debugger situation on linux is a horrible mess at the moment. There are some really good, also paid options, but those aren't nearly as mature as Remedy, nor are they as cheap. Honestly, there's likely quite a number of people who'd be willing to port Remedy to linux, all you'd have to do is ask.

**Comment by hsnyder** [comment]
+1 to the above. I would volunteer some time to help with the port, if it would help make it happen. I'd also sign an NDA.

**Comment by marvhus** [comment]
> +1 to the above. I would volunteer some time to help with the port, if it would help make it happen. I'd also sign an NDA.

I would also be willing to volunteer some time if that means we would get a Linux Version

**Comment by kdkopx** [comment]
Same as the others, I'd also definitely buy this if I could use it on Linux. I've been looking for a nice debugger like this for so long, but all of them just suck and in my experience make development much more annoying lot of the time. From what I seen online, this looks like what I want and I assure you, I am more than willing to give you 30 bucks. Good luck and happy hacking!

**Comment by Devastus** [comment]
Just adding my voice here, the moment RemedyBG would get a Linux version would be a 100% guaranteed instant buy. The only reason I haven't already done so yet is because I generally mostly use Linux, rarely do I use Windows especially for any coding tasks.

I most likely will end up buying it regardless at some point (if I happen to do more serious gamedev on Windows some day), but for my personal usage Linux support would be a monumental feature.

**Comment by 0xd61** [comment]
+1 for a  Linux Version. In the meantime, I can recommend the GDB frontend by nakst: https://github.com/nakst/gf. But not having to deal with GDB would be awesome!

**Comment by kuripa** [comment]
+1 Linux support. Working full time Linux now and really missing Remedybg. Would buy another License if this is happening 🫡

**Comment by Ddystopia** [comment]
I saw the features of this debugger and was shocked! I'm really looking forward to the port to Linux, even if the license costs 10 times as much.

**Comment by igaryhe** [comment]
i use windows at work and use linux at home. the only thing i miss on windows is remedybg. really wish we could have this available on linux. i would also love to pay again!

**Comment by mwcz** [comment]
Another +1 from me.  :dollar:

**Comment by kw3rk** [comment]
I've already paid, and would pay again for this. +1 from me.

**Comment by sayakbiswas** [comment]
Another vote from me!

**Comment by Dom324** [comment]
This issue is by far the most popular request considered both number of comments and number of upvotes.
@x13pixels could we get an update whether you have reconsidered this? :slightly_smiling_face: 

![image](https://github.com/x13pixels/remedybg-issues/assets/57506340/983daaab-4f65-48f6-a82f-8dd7a8d9500e)
![image](https://github.com/x13pixels/remedybg-issues/assets/57506340/8624907a-609a-494e-ba74-3e7e47324ce1)

**Comment by tdgroot** [comment]
With gdb 14.1 now supporting debug adapter protocol, does linux support for remedybg become more likely?

**Comment by krux02** [comment]
I just want to leave my comment here to mention that Linux support is still very desirable. Pretty please 🥺

**Comment by davidhaley** [comment]
+1

**Comment by derek57** [comment]
As requested by several GitHub users beforehand, I'd also appreciate to see this run within Linux.

**Comment by Siiiimon** [comment]
+1

**Comment by djmittens** [comment]
💯

**Comment by vereena0x13** [comment]
+1

**Comment by thisisnotcamilo** [comment]
+1

**Comment by HerrNamenlos123** [comment]
+1
I would love to pay for a product that i know is actually good, and currently I don't really have any other tools on Linux that even come close to a proper debugger. Currently the best bet is clangd/llvm in VSCode, which is still terrible. 

Yes, Linux People may be a bit hesitant to pay for software in general, but it's mainly because most paid software is from big corporations selling their data, and also because most software you pay for is actually still as crappy as free software. I think this is different with RemedyBg as it is well known for actually being of excellent quality, and it's also not from a big corp.

Very happy to pay and test it once it's available for Linux.

**Comment by saccarosium** [comment]
+1

**Comment by rakivo** [comment]
+1

**Comment by alaa-alawi** [comment]
+1

**Comment by ghost** [comment]
+1

**Comment by pprettysimpple** [comment]
+1

**Comment by ebenezerrahul** [comment]
+1

**Comment by cyberkm** [comment]
+1
